### PR TITLE
[MLIR][LLVM] DI Recursive Type fix for recursion via scope of composites

### PR DESCRIPTION
**Pull request by zyx-billy**
Fixes this bug for the previous recursive DI type PR: https://github.com/llvm/llvm-project/pull/80251#issuecomment-2007254788.

Drawing inspiration from how clang uses DIBuilder to build forward decls, this PR changes how placeholders are created & updated. Instead of requiring each recursive DIType to do in-place mutation, we simply ask for a temporary node as the placeholder, and run RAUW at the end when the concrete node is translated.

This has the side effect of simplifying what's needed to add recursion support for a type. Now only one additional method needs to be created for exporting. Concretely, for this PR, `translateImpl` for DICompositeType is back to the state it was before the previous PR, and the only net addition for DICompositeType is `translateTemporaryImpl`.

### CODE REVIEW

File: mlir/test/Target/LLVMIR/llvmir-debug.mlir [hunk 1]
```diff
@@ -402,3 +402,24 @@ llvm.func @func_debug_directives() {
...
+
+// -----
+
+// Test recursive composite type with recursion via its scope is translated.
+
+#di_composite_type_self = #llvm.di_composite_type<tag = DW_TAG_null, recId = distinct[0]<>>
+#di_file = #llvm.di_file<"test.mlir" in "/">
+#di_subroutine_type = #llvm.di_subroutine_type<types = #di_composite_type_self>
+#di_subprogram = #llvm.di_subprogram<scope = #di_file, name = "my_func", file = #di_file, line = 182, scopeLine = 182, subprogramFlags = Optimized, type = #di_subroutine_type>
+// The composite type whose scope leads to a self-recursive cycle.
```
Comment (gysit) [inline mlir/test/Target/LLVMIR/llvmir-debug.mlir:414]:
```suggestion
```

File: mlir/test/Target/LLVMIR/llvmir-debug.mlir [hunk 2]
```diff
@@ -402,3 +402,24 @@ llvm.func @func_debug_directives() {
 llvm.func @class_method() {
   llvm.return loc(#loc3)
 } loc(#loc3)
+
+// -----
+
+// Test recursive composite type with recursion via its scope is translated.
```
Comment (gysit) [inline mlir/test/Target/LLVMIR/llvmir-debug.mlir:408]:
```suggestion
// Ensures composite types with a recursive scope work.
```

**Review by gysit — APPROVED** [review]
Thanks, LGTM!

I did not know that particular API but that is a nice simplification. If you see some opportunities to strip and/or format the test case a bit that would be great.

**Comment by zyx-billy** [comment]
This windows build is taking forever... I'll check back tonight too but feel free to merge it if you need it sooner.

**Pull request merged**